python/lsst/ctrl/bps/clustered_quantum_graph.py

```diff
@@ -23,65 +23,473 @@
+        clusters : `Iterator`
+            Iterator over successors of given cluster.
+        """
+        return map(self.get_cluster, self._cluster_graph.successors(name))
+
+    def predecessors(self, name):
+        """Return clusters that are predecessors of the cluster
+        with the given name.
+
+        Parameters
+        ----------
+        name : `str`
+            Name of cluster for which need the predecessors.
+
+        Returns
+        -------
+        clusters : `Iterator`
+            Iterator over predecessors of given cluster.
+        """
+        return map(self.get_cluster, self._cluster_graph.predecessors(name))
+
+    def add_dependency(self, parent, child):
+        """Add a directed dependency between a parent cluster and a child
+           cluster.
+
+        Parameters
+        ----------
+        parent : `str` or `QuantaCluster`
+            Parent cluster.
+        child : `str` or `QuantaCluster`
+            Child cluster.
+
+        Raises
+        ------
+        KeyError
+            Raised if either the parent or child doesn't exist in the
+            ClusteredQuantumGraph.
+        """
+        if not self._cluster_graph.has_node(parent):
+            raise KeyError(f"{self.name} does not have a cluster named {parent}")
+        if not self._cluster_graph.has_node(child):
+            raise KeyError(f"{self.name} does not have a cluster named {child}")
+        _LOG.debug(f"add_dependency: adding edge {parent} {child}")
+
+        if isinstance(parent, QuantaCluster):
+            pname = parent.name
         else:
-            raise RuntimeError("Missing qgraph_node_ids attribute.")
+            pname = parent
+
+        if isinstance(child, QuantaCluster):
+            cname = child.name
+        else:
+            cname = child
+        self._cluster_graph.add_edge(pname, cname)
+
+    def __contains__(self, name):
+        """Check if a cluster with given name is in this ClusteredQuantumGraph.
+
+        Parameters
+        ----------
+        name : `str`
+            Name of cluster to check.
+
+        Returns
+        -------
+        found : `bool`
+            Whether a cluster with given name is in this ClusteredQuantumGraph.
+        """
+        return self._cluster_graph.has_node(name)
+
+    def save(self, filename, format_=None):
+        """Save the ClusteredQuantumGraph in a format that is loadable.
+        The QuantumGraph is saved separately if hasn't already been
+        serialized.
+
+        Parameters
+        ----------
+        filename : `str`
+            File to which the ClusteredQuantumGraph should be serialized.
+
+        format_ : `str`, optional
+            Format in which to write the data. It defaults to pickle format.
+        """
+        path = Path(filename)
+
+        # if format is None, try extension
+        if format_ is None:
+            format_ = path.suffix[1:]  # suffix includes the leading period
+
+        if format_ not in ("pickle"):
```

Comment:
I may be wrong, but judging from the error message which will be displayed if the conditions is true, you want to check for membership rather than substring matching. That is the condition should read `format_ not in ("pickle",)` (note the comma). Though probably even better approach would be using `set`, i.e., `format_ not in {"pickle"}`.

Reply:
Just FYI, if I'm correct, line 485 also needs to be fixed.